compiler/src/iree/compiler/Dialect/Flow/Transforms/test/set_encoding.mlir

```diff
@@ -913,3 +913,66 @@ util.func public @multi_batch_dim_generic(%arg0 : tensor<4x8x256x128xf32>, %arg1
+  %7 = flow.dispatch.region -> (tensor<?x?x?xi32>{%3, %0, %4}) {
+    %9 = tensor.empty(%3, %0, %4) : tensor<?x?x?xi32>
+    %c0_i32_0 = arith.constant 0 : i32
+    %10 = tensor.empty(%3, %0, %1) : tensor<?x?x?xi32>
+    %11 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel", "parallel"]} ins(%2 : tensor<?x?xi8>) outs(%10 : tensor<?x?x?xi32>) {
+    ^bb0(%in: i8, %out: i32):
+      %14 = arith.extui %in : i8 to i32
+      linalg.yield %14 : i32
+    } -> tensor<?x?x?xi32>
+    %12 = linalg.fill ins(%c0_i32_0 : i32) outs(%9 : tensor<?x?x?xi32>) -> tensor<?x?x?xi32>
+    %13 = linalg.batch_matmul_transpose_b ins(%11, %6 : tensor<?x?x?xi32>, tensor<?x?x?xi32>) outs(%12 : tensor<?x?x?xi32>) -> tensor<?x?x?xi32>
+    flow.return %13 : tensor<?x?x?xi32>
+  }
+  util.return %7 : tensor<?x?x?xi32>
+}
+// CHECK-DAG:  #[[MAP:.+]] = affine_map<(d0, d1, d2) -> (d1, d2)>
+// CHECK-DAG:  #[[MAP1:.+]] = affine_map<(d0, d1, d2) -> (d0, d1, d2)>
+// CHECK-DAG:  #[[MAP2:.+]] = affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>
+// CHECK-DAG:  #[[MAP3:.+]] = affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>
+// CHECK-DAG:  #[[MAP4:.+]] = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
+// CHECK:      util.func public @broadcasting_dequant_op(
+// CHECK:      %{{.+}} = flow.dispatch.region
+// CHECK:        %[[BCAST:.+]] = linalg.generic
+// CHECK:        %[[LHS:.+]] = iree_encoding.set_encoding %[[BCAST]] : tensor<?x?x?xi32>
+// CHECK-SAME:    -> tensor<?x?x?xi32, #iree_encoding.encoding
+// CHECK-SAME:      operand_index = 0 : index
+// CHECK-SAME:      element_types = [i32, i32, i32]
+// CHECK-SAME:      original_type = tensor<?x?x?xi32>
+// CHECK-SAME:      user_indexing_maps = [#[[MAP2]], #[[MAP3]], #[[MAP4]]]
+// CHECK-SAME:      round_dims_to = array<i64: 32, 32, 32>
+// CHECK:        %[[RHS:.+]] = iree_encoding.set_encoding %{{.+}} : tensor<?x?x?xi32>
+// CHECK-SAME:      operand_index = 1 : index
+// CHECK-SAME:      element_types = [i32, i32, i32]
+// CHECK-SAME:      original_type = tensor<?x?x?xi32>
+// CHECK-SAME:      user_indexing_maps = [#[[MAP2]], #[[MAP3]], #[[MAP4]]]
+// CHECK-SAME:      round_dims_to = array<i64: 32, 32, 32>
+// CHECK:        %[[INIT:.+]] = tensor.empty({{.+}}) : tensor<?x?x?xi32, #iree_encoding.encoding
+// CHECK-SAME:      operand_index = 2 : index
+// CHECK-SAME:      element_types = [i32, i32, i32]
+// CHECK-SAME:      original_type = tensor<?x?x?xi32>
+// CHECK-SAME:      user_indexing_maps = [#[[MAP2]], #[[MAP3]], #[[MAP4]]]
+// CHECK-SAME:      round_dims_to = array<i64: 32, 32, 32>
+// CHECK:        %[[FILL:.+]] = linalg.fill ins({{.+}}) outs(%[[INIT]]
+// CHECK:        %[[GEMM:.+]] = linalg.batch_matmul_transpose_b
+// CHECK-SAME:     ins(%[[LHS]], %[[RHS]]
+// CHECK-SAME:    outs(%[[FILL]]
+// CHECK:        %[[UNSET:.+]] = iree_encoding.unset_encoding %[[GEMM]]
+// CHECK:        %[[SLICE:.+]] = tensor.extract_slice %[[UNSET]]
+// CHECK:        flow.return %[[SLICE]]
```

Comment:
Just in case if others could miss the new test. This is what we expect when we set encodings after dispatch formation.

Reply:
but the set encoding is within the dispatch? So a later pass will split these apart?

Reply:
Yes, that's the plan. Max will send other PRs that hoist the set_encoding ops out, and fuse them with producer dispatches (or create a single dispatch for the op).

I think https://github.com/Max191/iree/commit/89722546c503c1034b6eb8c73f2b77b4c95f4bf3 is the prototype. Max could have a branch that has everything connected somewhere. Max, can you share the IR dumps with your full prototype? I think it is okay if it fails in materialization pass. Mahesh and I can take a look at IRs to get better ideas about what's going on at Flow level.

Reply:
Yes here is the branch for the prototype: https://github.com/Max191/iree/tree/data-tiling-fusion-prototype

And here are some IR dumps for a quantized matmul case: https://gist.github.com/Max191/d92c8750e6f3890af73512b84f93d4b6

We will have 2 new passes that hoist encoding ops out of dispatch regions and then fuse them into the appropriate dispatches (or form new dispatches).